File: src/main/scala/g0101_0200/s0139_word_break/Solution.scala
package g0101_0200.s0139_word_break

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table
// #Dynamic_Programming #Trie #Memoization #Algorithm_II_Day_15_Dynamic_Programming
// #Dynamic_Programming_I_Day_9 #Udemy_Dynamic_Programming #Big_O_Time_O(M+max*N)_Space_O(M+N+max)
// #2023_11_04_Time_482_ms_(66.67%)_Space_57.5_MB_(6.67%)

object Solution {
    def wordBreak(s: String, wordDict: List[String]): Boolean = {
        val dp = (0 until s.length() + 1).map(_ => false).toBuffer
        dp(0) = true
        val wordSet = wordDict.toSet

        def loop(j: Int, end: Int): Int = {
            if (j < end && dp(j) && wordSet.contains(s.substring(j, end))) {
                dp(end) = true
                end
            }
            else if (j < end) loop(j + 1, end)
            else -1
        }

        for (i <- 1 until s.length() + 1) {
            loop(0, i)
        }
        dp.last
    }
}

File: src/main/swift/g0101_0200/s0139_word_break/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table
// #Dynamic_Programming #Trie #Memoization #Algorithm_II_Day_15_Dynamic_Programming
// #Dynamic_Programming_I_Day_9 #Udemy_Dynamic_Programming #Top_Interview_150_1D_DP
// #Big_O_Time_O(M+max*N)_Space_O(M+N+max) #2024_06_27_Time_3_ms_(95.73%)_Space_16.4_MB_(60.50%)

class Solution {
    func wordBreak(_ s: String, _ wordDict: [String]) -> Bool {
        var words = [[Character]]()
        for word in wordDict {
            words.append(Array(word))
        }
        let sSlice = Array(s)[...]
        var visited = Array(repeating: false, count: sSlice.count)
        return canBeBroken(sSlice, words, &visited)
    }

    func canBeBroken(
        _ chars: ArraySlice<Character>, 
        _ words: [[Character]],
        _ visited: inout [Bool]
    ) -> Bool {
        let startIndex = chars.startIndex
        guard startIndex != chars.endIndex else { return true }
        guard !visited[startIndex] else { return false }
        for word in words {
            guard 
                chars.starts(with: word),
                canBeBroken(chars[startIndex.advanced(by: word.count)...], words, &visited)
            else { 
                continue 
            }
            return true
        }
        visited[startIndex] = true
        return false
    }
}
